drivers/modem/gsm_ppp.c

```diff
@@ -492,12 +630,58 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 		return;
 	}
 
+	/* Attached, clear retry counter */
+	gsm->attached = true;
+	gsm->attach_retries = 0;
+
+	LOG_DBG("modem attach returned %d, %s", ret, "read RSSI");
+	gsm->rssi_retries = GSM_RSSI_RETRIES;
+
+ attached:
+
+	/* Read connection quality (RSSI) before PPP carrier is ON */
+
+#if defined(CONFIG_MODEM_GSM_ENABLE_CESQ_RSSI)
+		ret = modem_cmd_send_nolock(&gsm->context.iface,
+				    &gsm->context.cmd_handler,
+				    &read_rssi_cmd, 1, "AT+CESQ",
+				    &gsm->sem_response, GSM_CMD_SETUP_TIMEOUT);
+#else
+		ret = modem_cmd_send_nolock(&gsm->context.iface,
+				    &gsm->context.cmd_handler,
+				    &read_rssi_cmd, 1, "AT+CSQ",
+				    &gsm->sem_response, GSM_CMD_SETUP_TIMEOUT);
+#endif
+
+		if (ret < 0) {
+			LOG_DBG("No answer to RSSI readout, %s", "ignoring...");
+		}
+
+		if (!(gsm->context.data_rssi && gsm->context.data_rssi != GSM_RSSI_INVALID &&
+		gsm->context.data_rssi < GSM_RSSI_MAXVAL)) {
+
+			LOG_DBG("Not valid RSSI, %s", "retrying...");
+			if (gsm->rssi_retries-- > 0) {
+				(void)k_work_reschedule(&gsm->gsm_configure_work,
+							K_MSEC(GSM_RSSI_RETRY_DELAY_MSEC));
+				return;
+			}
+		}
+
 #if defined(CONFIG_MODEM_CELL_INFO)
-	(void)gsm_query_cellinfo(gsm);
+
+		int counter = 0;
+
+		ret = -EINVAL;
+		while (counter < GSM_CELL_INFO_MAX_RETRIES && ret < 0) {
+			ret = gsm_query_cellinfo(gsm);
+			counter++;
+		}
```

Comment:
Another possibility would be to read cell info in rssi_handler(). Since cell info can change during the connection, this would be an advantage. The ublox driver is implementing it like this, and by calling the handler once during initialisation it also avoids some code duplication.

Reply:
addressed now, I have considered both MUX and without MUX possibilities.